src/installation_metadata.c

```diff
@@ -0,0 +1,180 @@
+#include <stdlib.h>
+#include <unistd.h>
+#include <postgres.h>
+#include <fmgr.h>
+#include <catalog/pg_type.h>
+#include <access/htup_details.h>
+#include <utils/builtins.h>
+#include <utils/fmgroids.h>
+#include <utils/lsyscache.h>
+#include <utils/datum.h>
+
+#include "catalog.h"
+#include "installation_metadata.h"
+#include "scanner.h"
+
+static Datum
+convert_value_to_text(Datum value, Oid from_type)
+{
+	bool		value_is_varlena;
+	Oid			value_out;
+
+	getTypeOutputInfo(from_type, &value_out, &value_is_varlena);
+
+	if (!OidIsValid(value_out))
+		elog(ERROR, "no output function for type %u", from_type);
+
+	return DirectFunctionCall1(textin, OidFunctionCall1(value_out, value));
+}
+
+static Datum
+convert_key_to_name(Datum value, Oid from_type)
+{
+	bool		value_is_varlena;
+	Oid			value_out;
+
+	getTypeOutputInfo(from_type, &value_out, &value_is_varlena);
+
+	if (!OidIsValid(value_out))
+		elog(ERROR, "no output function for type %u", from_type);
+
+	return DirectFunctionCall1(namein, OidFunctionCall1(value_out, value));
+}
+
+static Datum
+convert_text_to_value(Datum value, Oid to_type)
+{
+	Oid			value_in;
+	Oid			value_ioparam;
+
+	getTypeInputInfo(to_type, &value_in, &value_ioparam);
+
+	if (!OidIsValid(value_in))
+		elog(ERROR, "no input function for value type %u", to_type);
+
+	value = OidFunctionCall3(value_in,
+							 CStringGetDatum(TextDatumGetCString(value)),
+							 ObjectIdGetDatum(InvalidOid),
+							 Int32GetDatum(-1));
+	return value;
+}
+
+typedef struct DatumValue
+{
+	Datum		value;
+	Oid			typeid;
+	bool		isnull;
+} DatumValue;
+
+static bool
+installation_metadata_tuple_get_value(TupleInfo *ti, void *data)
+{
+	DatumValue *dv = data;
+
+	dv->value = heap_getattr(ti->tuple, Anum_installation_metadata_value, ti->desc, &dv->isnull);
+
+	if (!dv->isnull)
+		dv->value = convert_text_to_value(dv->value, dv->typeid);
+
+	return false;
+}
+
+static Datum
+installation_metadata_get_value_internal(Datum metadata_key,
+										 Oid key_type,
+										 Oid value_type,
+										 bool *isnull,
+										 LOCKMODE lockmode)
+{
+	Oid			key_out;
+	bool		key_is_varlena;
+	ScanKeyData scankey[1];
+	DatumValue	dv = {
+		.typeid = value_type,
+		.isnull = true,
+	};
+	Catalog    *catalog = catalog_get();
+	ScannerCtx	scanctx = {
+		.table = catalog->tables[INSTALLATION_METADATA].id,
+		.index = CATALOG_INDEX(catalog, INSTALLATION_METADATA, INSTALLATION_METADATA_PKEY_IDX),
+		.nkeys = 1,
+		.scankey = scankey,
+		.tuple_found = installation_metadata_tuple_get_value,
+		.data = &dv,
+		.lockmode = lockmode,
+		.scandirection = ForwardScanDirection,
+	};
+
+	getTypeOutputInfo(key_type, &key_out, &key_is_varlena);
+
+	if (!OidIsValid(key_out))
+		elog(ERROR, "no output function for key type %u", key_type);
```

Comment:
Again, some sort of prefix for our errors would be good, otherwise folks will find them in the logs and get very confused.